transcom/mymove
# Use asdf for node

## Use asdf for node

> **Warning**
> It is a-okay to say "Brandon, its not the time for this kind of thing" and we can close this PR. 

I ran this by folks in the BE/FE meeting on Tuesday Jul 18th, to which consensus was:
<img width="211" alt="image" src="https://github.com/transcom/mymove/assets/15805554/95d06e47-38d6-4b4f-b483-504a70217135">



## Summary

While setting up the project for the first time, it felt odd to use asdf for go but not for node, since it is capable of handling both. I'm sure there are historical reasons for the split (perhaps asdf did not have a nodejs plugin), etc.

> Is there anything you would like reviewers to give additional scrutiny?

- Yes! I'm completely new to the project, there could be unforeseen effects of this change. Please make me aware of the ones you are worried about if any!

- Companion PR(s)
  - https://github.com/transcom/mymove-docs/pull/386

Backwards compatibility:
- `.node-version`:
  - Folks who are using nodenv should be able to continue to use nodeenv. That means keeping the .node-version file around.
  - It is, in my opinion, not a good idea to force everyone to reconfigure their version manager right now. But the change will be socialized so that folks can swap on their own schedules.
- `.go-version`:
  - I'm new to the go ecosystem, so I'm not sure of the importance of this file. We currently use it explicitly in two places:
    - In the review app Dockerfile: https://github.com/transcom/mymove/blob/4098258c65e383e3540732dc2a23ee66c3f3bf9d/Dockerfile.reviewapp#L40 
    - In CircleCI config.yml as a [cache key](https://circleci.com/docs/caching/#restoring-cache): https://github.com/transcom/mymove/blob/d8424e0ebe41a672fe9b7a78dd6d1bd8be11b529/.circleci/config.yml#L479
      - For the purposes of a checksum, this could be updated to point to .tool-versions, if anyone has a strong opinion either way 

## Verification Steps for the Author

These are to be checked by the author.

I setup my environment with asdf for node. Initially, I allowed our scripts to install nodenv, but realized managing two different tools (`asdf` and `nodenv`) for node versions can cause some tricky to debug issues. So I went down the rabbit hole of exploring how difficult it would be to use asdf and run the project that way (+ update documentation and scripting). Turns out, not that bad!
- [x] Tested setting up project fresh with asdf instead of nodenv
- [x] Tested make targets relating to node version
- [x] Tested updated install script which relies on asdf to handle installation from .tool-versions, instead of the previous iteration of the script which was trying to circumvent asdf by parsing .tool-versions and then instructing asdf.

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

This is easiest to test when setting up the project from scratch. Running the make commands from [project onboarding documentation](https://transcom.github.io/mymove-docs/docs/about/application-setup/setup-manually/prerequisites)

1. run `make prereqs`
2. run `make deps`
3. ensure `make check_node_version` still works
   ```
   % rm -rf .check_node_version.stamp
   % make check_node_version
   ```
